docs(specs): Add `probe` as value to `startup_error_behavior`

LandonTClipp opened the pull request:
## Summary

This PR submits a spec for a new `probe` value to `startup_error_behavior`.

## Checklist
<!-- Mandatory
Please confirm the following by replacing the space with an "x" between the []:
-->

- [x] No AI generated code was used in this PR

## Related issues
<!-- Mandatory
All PRs should resolve an issue, if one does not exist, please open one.
-->

resolves #16028

LandonTClipp commented:
@srebhan sorry for the ping, just curious if you have time to give thoughts on this spec?

LandonTClipp commented on docs/specs/tsd-008-probe-on-startup.md:
```diff
@@ -0,0 +1,55 @@
+
+## Configuration
+
+The Telegraf project has already introduced the `startup_error_behavior` configuration which allows the user to define how Telegraf should behave if the plugin fails to start. The `ignore` value allows the plugin to be ignored if the `Start()`/`Connect()` method returns an error. We propose to add an additional value to this parameter called `probe`, that will behave as a superset of the `ignore` ...
```
I'm not sure if I totally understand the question. Is your question asking if we can just enable probing for any plugin that implements it, without needing a new config parameter? Personally, I tend to lean towards the end of giving the user _more_ control over what happens, not less. In the case of output plugins, it may very well be the case that users emphatically don't want the plugin probing on startup because that means, probably, that you're sending some fake metrics out somewhere.

LandonTClipp commented on docs/specs/tsd-008-probe-on-startup.md:
````diff
@@ -0,0 +1,55 @@
+}
+```
+
+In almost all respects, a `Probe` is similar to a `Gather`, the only difference being that `Probe` does not record any metrics and only returns information on the success or failure of the action.
````
Changed the wording, let me know your thoughts.

LandonTClipp commented:
@srebhan @DStrand1 I addressed all of your comments, including the line width comment. Thank you for your thoughtful insight 🙏🏻

srebhan commented on docs/specs/tsd-008-probe-on-startup.md:
```diff
@@ -0,0 +1,88 @@
+2. Return an error if information cannot be gathered (in the case of inputs) or 
+sent (in the case of outputs) due to unrecoverable issues. For example, invalid 
+authentication, missing permissions, or non-existent endpoints.
+3. Otherwise, return `nil`.
```
```suggestion
3. Otherwise, return `nil` indicating the plugin will be fully functional.
```

srebhan commented on docs/specs/tsd-008-probe-on-startup.md:
```diff
@@ -0,0 +1,88 @@
+authentication, missing permissions, or non-existent endpoints.
+3. Otherwise, return `nil`.
+
+## Configuration
+
+The Telegraf project has already introduced the `startup_error_behavior` 
+configuration which allows the user to define how Telegraf should behave if the 
+plugin fails to start. The `ignore` value allows the plugin to be ignored if the 
+`Start()`/`Connect()` method returns an error. We propose to add an additional 
+value to this parameter called `probe`, that will behave as a superset of the 
+`ignore` behavior. When `startup_error_behavior=probe`, Telegraf will perform 
+the following steps on plugin startup:
+
+1. Check if the plugin implements `ProbePlugin`. If it does not, Telegraf will 
+fatally exit with a log message indicating that the supplied configuration is 
+invalid for the plugin.
+2. Call the `Start()` or `Connect()` method.
+3. If an error is returned, cause the plugin to be ignored in the same manner as 
+if `ignore` was specified.
+4. Call `Probe()`. If `Probe()` returns an error, cause the plugin to be ignored.
```
Remove as this is already covered in the other spec (TSD-006).

srebhan commented on docs/specs/tsd-008-probe-on-startup.md:
```diff
@@ -0,0 +1,88 @@
+
+## Plugin Requirements
+
+As already stated, plugins participating in the `probe` scheme must implement 
+the `ProbePlugin` interface. The exact way the plugin implements the behavior 
+will depend on the plugin in question, but generally it should take the same 
+actions as it would with `Gather()` or `Write()`, such that failures during 
+`Gather()`/`Write()` would also imply a failure of `Probe()`. If `Probe()` 
+returns an error, the plugin's `Close()` method should still be safe to call.
+
+It should be noted that for output plugins, it's advisable to implement 
+`Probe()` in a way that doesn't write real metrics to the backend if possible. 
+How this might be done depends on the output in question.
+
+Plugins should return `nil` upon successful probes or a retriable error (via 
+predefined error type) to enable the defined ignoring behavior. A non-retriable 
+error (via predefined error type) or a generic error will bypass the startup 
+error behaviors and Telegraf must fail and exit in the startup phase.
```
```suggestion
Plugins that allow probing must implement  the `ProbePlugin` interface. The
exact implementation depend on the plugin's functionality and requirements,
but generally it should take the same actions as it would during normal operation
e.g. calling `Gather()` or `Write()` and check if errors occur. If probing fails, it must
be safe to call the plugin's `Close()` method.

Input plugins must *not* produce metrics, output plugins must *not* send any
metrics to the service. Plugins must *not* influence the later data processing or
collection by modifying the internal state of the plugin or the external state of the
service or hardware. For example, file-offsets or other service states must be
reset to not loose data during the first gather or write cycle.

Plugins must return `nil` upon successful probing or an error otherwise.
```

srebhan reviewed (commented):
Thanks @LandonTClipp for working with us on this spec! Some more comments from my side, but I feel we are very close...

LandonTClipp commented on docs/specs/tsd-008-probe-on-startup.md:
```diff
@@ -0,0 +1,88 @@
+
+## Probing
+
+First, we must define what it means for a plugin to `Probe`. Probing is an 
+action whereby the plugin will attempt to communicate with its external service, 
+device, entity, or executable as if it were gathering real metrics. The payload 
+of this communication attempt is not used to record any metrics. 
```
I like it!

LandonTClipp commented:
@srebhan addressed comments and made some typo fixes.

Also related, what do you think about reverting https://github.com/influxdata/telegraf/pull/15916 in favor of this spec?

srebhan reviewed (commented):
Thanks @LandonTClipp! The PR looks great! Could you please rename the file to `tsd-009-probing.md` as we already merged TSD-008 in the meantime. Please also change the reference in TSD-006 accordingly!

Regarding reverting the code in `nvidia-smi`, yes I think you can revert _or_ change the implementation to reflect this spec as we did not release the mentioned feature yet...

LandonTClipp commented:
> Regarding reverting the code in nvidia-smi, yes I think you can revert or change the implementation to reflect this spec as we did not release the mentioned feature yet...

Excellent, I'll probably just revert for now, in case I cannot implement this spec in time for the release.

srebhan reviewed (approved):
Thanks a lot @LandonTClipp!

LandonTClipp commented:
@DStrand1 sorry for the ping but I wanted to check what you think of this. I'm motivated to implement it for the plugins I care about :D

LandonTClipp commented:
@srebhan @DStrand1 pinging again on this

srebhan commented:
@LandonTClipp could you please fix the linter issues? Seems like you do have a lot of trailing spaces in markdown....

LandonTClipp commented:
@srebhan done!

DStrand1 reviewed (approved):
Thanks for the great work!

Hipska commented on docs/specs/tsd-006-startup-error-behavior.md:
```diff
@@ -75,6 +75,19 @@ must *not* fail on startup errors and should continue running. On startup error,
+plugin must be completely removed from processing, similar to the `ignore`
+behavior. Additionally, Telegraf must probe the plugin (as defined in
+[TSD-009][tsd_009]) after startup, if it implements the `ProbePlugin` interface.
+If probing is available *and* returns an error Telegraf must *ignore* the
+plugin as-if it was not configured at all.
```
It is not clear what happens if the probing is not returning an error. As it reads that this probing only needs to be done when startup returns errors.

LandonTClipp replied:
@Hipska perhaps one more paragraph added here: https://github.com/influxdata/telegraf/pull/16052/files#diff-2d519c82d1022b0befbd7601817fd1b2073ad5e9b607cd9c2205f0529b7d0fffR47 more clearly outlining what Telegraf does on Probe success is what you're looking for? I'm happy to add more clarification.

Hipska replied:
Yeah indeed.

Also, is probing really only on startup? Meaning if conditions change and recourses come available, that will only take effect when telegraf gets restarted?

LandonTClipp replied:
That's correct. Sorry for the delayed response on the update to the spec, I will get to it soon.

This spec is backwards compatible so if you don't want this behavior, you don't need to do anything.

Hipska replied:
As said in other comment, maybe the spec should clarify a use case? When would it be handy to only be able to recover by restarting telegraf completely?

Hipska commented on docs/specs/tsd-009-probe-on-startup.md:
```diff
@@ -0,0 +1,68 @@
+
+## Overview
+
+When plugins are first instantiated, Telegraf will call the plugin's `Start()`
+method (for inputs) or `Connect()` (for outputs) which will initialize its
+configuration based off of config options and the running environment. It is
+sometimes the case that while the initialization step succeeds, the upstream
+service in which the plugin relies on is not actually running, or is not capable
+of being communicated with due to incorrect configuration or environmental
+problems. In situations like this, Telegraf does not detect that the plugin's
```
Initialisation happens in `Init()`, connection attempts in `Start()`/`Connect()`.

Hipska replied:
@LandonTClipp
